feat(kinesisfirehose): add HTTP Endpoint and Datadog destination

benjaminpottier opened the pull request:
Issue # (if applicable)

Closes #20354 and #15502

### Reason for this change

Adding support for HTTP destinations and adding  Datadog as an opinionated third-party destination.

### Description of changes

Adds support for using Datadog as a delivery destination for Amazon Data Firehose using an L2 HTTP endpoint destination.

### Describe any new or updated permissions being added

Nothing.

### Description of how you validated changes

Added unit tests. Integration tests (`integ.http-endpoint.ts` and `integ.datadog.ts`) are included and deployed with integ-runner.

### Checklist
- [ ] My code adheres to the [CONTRIBUTING GUIDE](https://github.com/aws/aws-cdk/blob/main/CONTRIBUTING.md) and [DESIGN GUIDELINES](https://github.com/aws/aws-cdk/blob/main/docs/DESIGN_GUIDELINES.md)

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation reviewed (changes requested):
(This review is outdated)

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 918ac17fb8084dd1b65f58bcc98aa745cc65d833
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=7630f584-9640-4282-844e-63bd138e075f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

benjaminpottier commented:
This is my first time creating a pull request for this project so any feedback is appreciated. I'm still figuring out the integration tests and have left this in draft while I work on it.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: dcc4a66c405bffcf6ee48b95d03e64c2c43cfe68
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=ad841cef-b4db-4daa-b493-4a8048d9085e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: b15adc8f78a33a8af4d58a57e2b3d7d30f132e9b
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=1da639b5-b24d-4bf3-9700-d9f63d4c3dff%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: fb1aad7453fc7ffe5339e00fd70023bc617ae8f1
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=3c2b76e9-b689-4ccf-a08c-842bbb79cfa0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: b3c4426d30c101c2807289dad11cbd14108a2f75
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0c39e0b2-d0b0-46c8-9d2e-3c929d6426c1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 6506e5ce0a320f0a63cc1696ec3892eed2f55d18
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a202927c-a47b-44d0-91e9-f85eb07c80fc%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 0acdb243894ea3056233d4135b1b3165d9e250ce
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8d6ff0f9-4867-480a-8be1-f7f0d4c03b35%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

benjaminpottier commented:
Seemed fairly obvious after a while that an L2 for an HTTP destination should come first. Not sure if this would be the right approach to reuse it for Datadog.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: ad7adc7a2168db702c4e4ac01dc6fccfb917192c
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=7ff67b2d-9c53-49da-9ca5-f83c1d8a59f6%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 20a8cbf52ebe765852c06f572a6cda71ac304cef
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=02c90d2f-ce65-4006-bb2f-8a6138dfdc87%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 4fa7a075cfa32471b5937d46e8376f3112e69aaf
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a8ee8dd7-a5df-49b6-b850-c3411f1f7515%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 0ca8c6cb135049d312866dbe04312797363c19f7
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c316253f-4083-497e-8d21-b54df5089a4a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: d1f31e356155493a1b5dfe5bfae1d9e9954a9913
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=90cd0dda-549c-4067-9f8f-c94c821e9463%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 5ffba59acc3e7878e20d45e050f41fcc6f6db48d
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=7321489a-535d-4d29-8110-52bd62f02061%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: a5f07dcb8c54eda2d709c3c147f653644f0d19cc
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=a9391ede-8cc2-42d0-ad37-ba019ca8a35a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: fb42a8dd31850e50384d4b08624915a7b1c9962b
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=f5fcebf0-982a-4a51-9365-2a022157d7fb%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: cdae4242490d09d710c747e7f9a9d6c5edfeb788
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4d44220d-e92e-4132-9039-a1b0f3b50198%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 846759d117e6b0c1ff0b46eab3d66a4a079bba6d
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2cbfdcee-3a25-4289-ac91-4b0a36a07a0f%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: daefa3f27dede7927ee06d3adbdb43c7d45ce8ef
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c8f12b1e-3191-4b3e-a716-648caa969621%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 079fb475f9271d9428c65591bb5d07d8eb191fb0
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=db580f67-a480-4b80-9b68-ef462d82137a%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: fac92a7ed844b5793c2fad80e4a2c08d08674aa0
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=542ee473-5009-4cdc-8180-af9add82efaf%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

alvazjor commented on packages/aws-cdk-lib/aws-kinesisfirehose/lib/datadog.ts:
```diff
@@ -0,0 +1,133 @@
+  /**
+   * Japan Location
+   */
+  DATADOG_LOGS_AP1 = 'https://aws-kinesis-http-intake.loqs.ap1.datadoqhq.com/api/v2/loqs?dd-protocol=aws-kinesis-firehose',
```
**Blocking**

This AP1 URL has three `g`→`q` typos: `loqs` / `datadoqhq` / `loqs`. As written, `datadoqhq.com` does not resolve, so customers selecting AP1 get silent delivery failures.

Compare the `DATADOG_LOGS_US3`/`US5` members just above, which use the correct `logs.<region>.datadoghq.com/api/v2/logs` form. The Datadog intake host is `datadoghq.com` ([AWS Big Data blog](https://aws.amazon.com/blogs/big-data/analyze-logs-with-datadog-using-amazon-kinesis-data-firehose-http-endpoint-delivery/)).

```ts
DATADOG_LOGS_AP1 = 'https://aws-kinesis-http-intake.logs.ap1.datadoghq.com/api/v2/logs?dd-protocol=aws-kinesis-firehose',
```

alvazjor commented on packages/aws-cdk-lib/aws-kinesisfirehose/lib/http-endpoint.ts:
```diff
@@ -0,0 +1,201 @@
+   * The access key required for Kinesis Firehose to authenticate with the HTTP endpoint selected as the destination.
+   * @default - None
+   */
+  readonly accessKey?: string;
```
**Blocking**

`accessKey` is the HTTP endpoint's authentication credential, so typing it as a plain `string` violates the CDK secrets rule — AGENTS.md: "You MUST use `cdk.SecretValue` for any prop that accepts a secret value." A bare `string` also invites hardcoding the key in source.

Let's type it correctly. Changing it to `cdk.SecretValue` matches the established pattern for endpoint credentials (e.g. CodePipeline `oauthToken`, CodeBuild `accessToken`), with the construct consuming it via `unsafeUnwrap()`:

```ts
// HttpEndpointConfig
readonly accessKey?: cdk.SecretValue;

// bind()
...this.props.endpointConfig.accessKey && { accessKey: this.props.endpointConfig.accessKey.unsafeUnwrap() },
```

alvazjor commented on packages/aws-cdk-lib/aws-kinesisfirehose/lib/datadog.ts:
```diff
@@ -0,0 +1,133 @@
+  /**
+   * The URL of the Datadog endpoint.
+   */
+  readonly url: DatadogLogsEndpointUrl | DatadogMetricsEndpointUrl | DatadogConfigurationsEndpointUrl;
```
**Blocking**

This is a TypeScript union of three enums. AGENTS.md Props Design says "No TypeScript union types (jsii incompatible) — use enum-like classes, separate props, or factory methods instead", and jsii will reject this in cross-language compilation.

Consider collapsing the three endpoint enums into one enum-like class so predefined members and custom URLs both work (same pattern as `ec2.InstanceType`):

```ts
export class DatadogEndpoint {
  public static readonly LOGS_US1 = new DatadogEndpoint('https://aws-kinesis-http-intake.logs.datadoghq.com/v1/input');
  // ... other members ...
  public static of(url: string): DatadogEndpoint { return new DatadogEndpoint(url); }
  private constructor(public readonly url: string) {}
}
```

alvazjor commented on packages/aws-cdk-lib/aws-kinesisfirehose/lib/http-endpoint.ts:
```diff
@@ -0,0 +1,201 @@
+          },
+        },
+        cloudWatchLoggingOptions: loggingOptions,
+        processingConfiguration: createProcessingConfig(scope, role, this.props.processor),
```
**Blocking**

When this PR was opened, `createProcessingConfig(scope, role, processor)` was valid. It has since been refactored to `createProcessingConfig(scope, props, options)` (`lib/private/helpers.ts:131`), so after the required merge from `main` this call won't compile. It also only forwards the deprecated `processor` and silently drops `processors`.

Passing the full props fixes both — the current helper reads `processor` and `processors` off `props` (see `s3-bucket.ts:167`):

```ts
processingConfiguration: createProcessingConfig(scope, this.props, { role }),
```

alvazjor commented on packages/aws-cdk-lib/aws-kinesisfirehose/lib/http-endpoint.ts:
```diff
@@ -0,0 +1,201 @@
+
+    const { backupConfig, dependables: backupDependables } = createBackupConfig(scope, role, {
+      mode: this.props.backupMode === HTTPBackupMode.ALL ? S3BackupMode.ALL : S3BackupMode.FAILED,
+    })!;
```
**Suggestion**

`HttpEndpointProps extends CommonDestinationProps`, so it inherits `s3Backup` (bucket, buffering, encryption, prefixes), but `bind()` never reads it — it always creates a new bucket from a derived `mode`. A user who sets `s3Backup: { bucket: myBucket }` sees nothing happen, which is a silent footgun and diverges from `S3Bucket`, which passes `this.props.s3Backup` straight through.

Could we either honor `s3Backup` (merging the user's bucket with the derived `mode`) or, if HTTP destinations must always own their backup bucket, not inherit a prop we discard? Honoring it also removes the need for the `!` non-null assertion here.

alvazjor commented on packages/aws-cdk-lib/aws-kinesisfirehose/test/http-endpoint.test.ts:
```diff
@@ -0,0 +1,53 @@
+import * as cdk from '../..';
+import { Match, Template } from '../../assertions';
+import * as iam from '../../aws-iam';
+import * as secretsmanager from '../../aws-secretsmanager';
```
**Nit**

`Match`, `iam`, and `secretsmanager` are imported but unused here (and `Match`/`iam` are unused in `datadog.test.ts`), which fails the repo's ESLint no-unused-vars rule and likely contributes to the red CI. Removing them — or, better, adding the coverage that would use them (a `secret` grant assertion, a user-provided `role`) — clears it.

alvazjor commented on packages/aws-cdk-lib/aws-kinesisfirehose/lib/http-endpoint.ts:
```diff
@@ -0,0 +1,201 @@
+/**
+ *  An HTTP destination for data from a Kinesis Data Firehose delivery stream.
+ */
+export class HTTPEndpoint implements IDestination {
```
**Suggestion**

CDK spells acronyms as words in PascalCase — AGENTS.md: "`JsonPattern` not `JSONPattern`" — so the convention here is `Http*` (cf. `apigatewayv2.HttpApi`, `cloudfront-origins.HttpOrigin`). This applies to `HTTPEndpoint`, `HTTPEndpointConfig`, `HTTPEndpointProps`, `HTTPAttribute`, `HTTPCompression`, `HTTPBackupMode`, `HTTPBufferingHints`, and `HTTPRetryOptions`. Since these are exported and effectively permanent once released, it's worth renaming to `Http*` before merge.

alvazjor commented on packages/aws-cdk-lib/aws-kinesisfirehose/lib/datadog.ts:
```diff
@@ -0,0 +1,133 @@
+      },
+      backupMode: props.backupMode ?? HTTPBackupMode.FAILED,
+      attributes: props.tags ?? [],
+    });
```
**Suggestion**

The `Datadog` wrapper hardcodes buffering (60s/4 MiB), retry (60s), and GZIP, and `DatadogProps` does not extend `CommonDestinationProps`, so customers can't set `role`, `processors`, `loggingConfig`, or their own backup bucket on a Datadog stream — capabilities the underlying `HttpEndpoint` already has. AGENTS.md asks L2s to "expose the full AWS service surface area — never omit capabilities. Provide sensible defaults users can override."

What do you think about having `DatadogProps extends CommonDestinationProps` and surfacing `bufferingHints`/`retryOptions`/`requestCompression` as optional overrides, keeping today's values as defaults?

alvazjor commented on packages/aws-cdk-lib/aws-kinesisfirehose/lib/http-endpoint.ts:
```diff
@@ -0,0 +1,201 @@
+          bufferingHints: {
+            ...this.props.bufferingHints.interval && { intervalInSeconds: this.props.bufferingHints.interval.toSeconds() },
+            ...this.props.bufferingHints.size && { sizeInMBs: this.props.bufferingHints.size.toMebibytes() },
+          },
```
**Suggestion**

The S3 path validates buffering via `createBufferingHints` (1-128 MiB, ≤900s, with `Token.isUnresolved` guards), but the HTTP path renders `bufferingHints`/`retryOptions` inline with no validation, so bad values fail only at deploy. Could we reuse/extend `createBufferingHints` and add a retry-duration bound (Firehose caps retries at 7200s), throwing `ValidationError` with a `lit` error code (matching the rest of the module) so misconfiguration fails fast at synth?

alvazjor commented on packages/aws-cdk-lib/aws-kinesisfirehose/lib/http-endpoint.ts:
```diff
@@ -0,0 +1,201 @@
+      mode: this.props.backupMode === HTTPBackupMode.ALL ? S3BackupMode.ALL : S3BackupMode.FAILED,
+    })!;
+
+    if (this.props.endpointConfig.secret) {
+      this.props.endpointConfig.secret.grantRead(role);
```
**Suggestion**

`accessKey` and `secret` are alternative authentication mechanisms for the HTTP endpoint — Firehose uses one or the other. Right now both can be set and both would render. Could we add a `ValidationError` when both are provided, so the conflict surfaces at synth with a clear message rather than producing an ambiguous template?

alvazjor commented:
Hey @benjaminpottier, checking in on this one since it's been quiet for a couple of weeks. Are you still able and interested in taking this one forward? Totally fine either way, just want to know whether to keep it on the active list or not.

If you do want to continue, the next steps are: address the change requests from the open review, merge the latest main into your branch to clear the conflicts (that should also get CI green again), and wrap up the integration tests. Happy to help if any of the feedback is unclear or you get stuck. And if you'd rather hand it off, just say the word, no problem at all.

benjaminpottier commented:
> Hey @benjaminpottier, checking in on this one since it's been quiet for a couple of weeks. Are you still able and interested in taking this one forward? Totally fine either way, just want to know whether to keep it on the active list or not.
> 
> If you do want to continue, the next steps are: address the change requests from the open review, merge the latest main into your branch to clear the conflicts (that should also get CI green again), and wrap up the integration tests. Happy to help if any of the feedback is unclear or you get stuck. And if you'd rather hand it off, just say the word, no problem at all.

Hi @alvazjor I will take a look at this soon. It had fallen off my radar.

benjaminpottier commented:
> > Hey @benjaminpottier, checking in on this one since it's been quiet for a couple of weeks. Are you still able and interested in taking this one forward? Totally fine either way, just want to know whether to keep it on the active list or not.
> > If you do want to continue, the next steps are: address the change requests from the open review, merge the latest main into your branch to clear the conflicts (that should also get CI green again), and wrap up the integration tests. Happy to help if any of the feedback is unclear or you get stuck. And if you'd rather hand it off, just say the word, no problem at all.
> 
> Hi @alvazjor I will take a look at this soon. It had fallen off my radar.

I think I addressed all your comments and I added more test coverage.

alvazjor commented on packages/aws-cdk-lib/aws-kinesisfirehose/lib/destination.ts:
```diff
@@ -11,6 +11,12 @@ export interface DestinationConfig {
    */
   readonly extendedS3DestinationConfiguration?: CfnDeliveryStream.ExtendedS3DestinationConfigurationProperty;
+  /**
+   * Datadog destination configuration properties.
```
This field is the generic HTTP endpoint destination config (both `HttpEndpoint` and `Datadog` render into it), but the doc line labels it as Datadog-specific. Reword the summary line to describe the HTTP endpoint, matching the S3 sibling above it ("S3 destination configuration properties").

Current:

```ts
/**
 * Datadog destination configuration properties.
 *
 * @default - HTTP destination is not used.
 */
readonly httpEndpointDestinationConfiguration?: CfnDeliveryStream.HttpEndpointDestinationConfigurationProperty;
```

Suggested:

```ts
/**
 * HTTP endpoint destination configuration properties.
 *
 * @default - HTTP destination is not used.
 */
readonly httpEndpointDestinationConfiguration?: CfnDeliveryStream.HttpEndpointDestinationConfigurationProperty;
```

alvazjor commented on packages/aws-cdk-lib/aws-kinesisfirehose/lib/datadog.ts:
```diff
@@ -0,0 +1,169 @@
+  /**
+   * Logs — AP1 (Japan)
+   */
+  public static readonly LOGS_AP1 = new DatadogEndpoint('https://aws-kinesis-http-intake.logs.ap1.datadoqhq.com/api/v2/logs?dd-protocol=aws-kinesis-firehose');
```
This still has the `datadoqhq` typo from the last round. The `logs.` segments are fixed now, but the host is `datadoqhq.com` (a `g` to `q` typo), which does not resolve, so AP1 customers get silent delivery failures. `METRICS_AP1` and `CONFIGURATION_AP1` just below use the correct `datadoghq.com`.

```ts
public static readonly LOGS_AP1 = new DatadogEndpoint('https://aws-kinesis-http-intake.logs.ap1.datadoghq.com/api/v2/logs?dd-protocol=aws-kinesis-firehose');
```

alvazjor commented on packages/aws-cdk-lib/aws-kinesisfirehose/README.md:
````diff
@@ -84,6 +84,31 @@ Amazon Data Firehose supports multiple AWS and third-party services as destinati
+
+```ts
+declare const endpointConfig: firehose.HTTPEndpointConfig;
+const httpDestination = new firehose.HTTPEndpoint({
````
`HTTPEndpointConfig` and `HTTPEndpoint` were renamed to `HttpEndpointConfig` / `HttpEndpoint`, so these no longer exist. README snippets are compiled by rosetta, so this breaks the build, and it also ships a broken example.

```ts
declare const endpointConfig: firehose.HttpEndpointConfig;
const httpDestination = new firehose.HttpEndpoint({
  endpointConfig,
});
```

alvazjor commented on packages/aws-cdk-lib/aws-kinesisfirehose/README.md:
```diff
@@ -84,6 +84,31 @@ Amazon Data Firehose supports multiple AWS and third-party services as destinati
+declare const apiKey: secretsmanager.Secret;
+const datadogDestination = new firehose.Datadog({
+  apiKey,
+  url: firehose.DatadogLogsEndpointUrl.DATADOG_LOGS_US1,
```
`DatadogLogsEndpointUrl` was removed and `Datadog` no longer takes a `url` prop; it takes `endpoint: DatadogEndpoint`. This breaks rosetta and is a broken example.

```ts
const datadogDestination = new firehose.Datadog({
  apiKey,
  endpoint: firehose.DatadogEndpoint.LOGS_US1,
});
```

alvazjor commented on packages/@aws-cdk-testing/framework-integ/test/aws-kinesisfirehose/test/integ.http-endpoint.ts:
```diff
@@ -0,0 +1,28 @@
+const stack = new cdk.Stack(app, 'aws-cdk-firehose-delivery-stream-s3-all-properties');
+
+const deliveryStream = new firehose.DeliveryStream(stack, 'DeliveryStream', {
+  destination: new firehose.HTTPEndpoint({
```
`firehose.HTTPEndpoint` was renamed to `HttpEndpoint`, so this integ test will not compile. Update it to `new firehose.HttpEndpoint({ ... })` and re-run integ-runner to regenerate the snapshot from the renamed source.

alvazjor commented on packages/@aws-cdk-testing/framework-integ/test/aws-kinesisfirehose/test/integ.datadog.ts:
```diff
@@ -0,0 +1,28 @@
+const deliveryStream = new firehose.DeliveryStream(stack, 'DeliveryStream', {
+  destination: new firehose.Datadog({
+    apiKey: Secret.fromSecretNameV2(stack, 'DatadogApiKey', 'DatadogApiKey'),
+    url: firehose.DatadogLogsEndpointUrl.DATADOG_LOGS_EU,
```
Rename issue: `Datadog` no longer takes `url` / `DatadogLogsEndpointUrl`; it takes `endpoint: DatadogEndpoint`. This will not compile. The rendered template should be unchanged since `LOGS_EU` resolves to the same URL as the old `DATADOG_LOGS_EU`.

```ts
destination: new firehose.Datadog({
  apiKey: Secret.fromSecretNameV2(stack, 'DatadogApiKey', 'DatadogApiKey'),
  endpoint: firehose.DatadogEndpoint.LOGS_EU,
}),
```

alvazjor commented on packages/aws-cdk-lib/aws-kinesisfirehose/test/datadog.test.ts:
```diff
@@ -0,0 +1,290 @@
+  });
+
+  describe('DatadogEndpoint', () => {
+    test.each([
```
This mapping table is thorough but is missing `LOGS_AP1`, `CONFIGURATION_AP1`, and `CONFIGURATION_US5`. `LOGS_AP1` is the member with the typo above, so adding it would have caught the bug. Could you add the three missing rows, and maybe assert every URL is well-formed (host ends in `datadoghq.com` / `.eu` / `ddog-gov.com`) so this class of typo fails a test rather than shipping?

alvazjor commented on packages/aws-cdk-lib/aws-kinesisfirehose/lib/http-endpoint.ts:
```diff
@@ -0,0 +1,241 @@
+      this.props.role ??
+			new iam.Role(scope, 'Http Destination Role', {
+			  assumedBy: new iam.ServicePrincipal('firehose.amazonaws.com'),
+			});
```
These lines use tabs while the rest of the file uses two-space indentation; ESLint will flag it (and it may be part of the red CI). `eslint --fix` clears it.

alvazjor commented:
@benjaminpottier The build failed, as exptected, due to the unchanged references from the first implementation. I think all is already in the review comments, just need to address them.

alvazjor commented:
Hey @benjaminpottier, checking in again, no pressure. Mostly want to know if you still have time for this one.

The last revision was a real step up. The accessKey to SecretValue change, the DatadogEndpoint class, the validation, and the extra unit tests all landed well.

What is left is small:
- Apply the HTTP* to Http* and DatadogLogsEndpointUrl to DatadogEndpoint rename in the README and the two integ test files (integ.http-endpoint.ts and integ.datadog.ts). That is what is failing the build. The library and unit tests were already renamed, these two spots were missed.
- Fix the LOGS_AP1 typo, the host reads datadoqhq instead of datadoghq.
- Merge main to clear the conflicts.
- Regenerate the integ snapshots after the rename.

The rest of the notes are in the existing review comments, so you do not have to re-read the whole thread.

If the integ tests are the part you are stuck on, I am still happy to run them and update the snapshots for you. Just say the word.

Either way, reply here so I know whether to keep this on the active list.

alvazjor commented:
Hi @benjaminpottier,

Thanks again for all the work on this one. Quick recap of where we are: you did a solid rework back on 06-29, I left the follow-up review on 06-30 with the last few items (the AP1 URL typo, and the README plus the two integ tests still referencing the old HTTP*/DatadogLogsEndpointUrl names), and I checked in again on 07-31. I have not heard back since, so I want to lay out what happens next.

This PR closes two popular requests (#15502 and #20354) and it is genuinely close. What is left is mechanical: fix the one-character datadoqhq typo, update the README examples to the renamed API, rename the two integ test sources, and re-run integ-runner so the snapshots regenerate. Because it is this close and the demand is real, I would like to finish it off and get it merged.

Here is the plan: if I do not hear from you by 08-10, I will push the remaining commits to your branch to wrap it up, keeping you as the author with Co-authored-by. If you would rather finish it yourself, just say so before then and I will hold off as long as you need. Either way is fine, and thanks for pushing this as far as you have.

alvazjor commented:
Hi @benjaminpottier,

Following up on my 08-07 note: since I did not hear back, I have pushed the remaining changes to your branch to wrap this up. You stay the author, my commit carries a Co-authored-by trailer so your credit is preserved, and your own commits are untouched.

What I pushed:
- Fixed the LOGS_AP1 host typo (datadoqhq to datadoghq).
- Finished the HTTP* to Http* and DatadogLogsEndpointUrl to DatadogEndpoint rename in the README and the two integ test sources. The library and unit tests were already renamed; these two spots were the ones left.
- Merged the latest main to clear the conflicts.

The build and the aws-kinesisfirehose unit tests pass locally. CI and the integration test deployment are running now, and I will regenerate the integ snapshots and take it through to merge. 

Thanks for pushing this as far as you did. It closes two long-standing requests, #15502 and #20354.
